File: go/aws/handlers.go
package aws

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"

	"github.com/aws/aws-lambda-go/events"

	"github.com/sns45/anyhook/go/core"
)

// tenantHeader is the trusted header carrying the caller's tenant id.
// Like the TS ingest handler (handlers.ts), anyhook does NOT authenticate end
// customers here (G8): mount this behind the producer's own auth, which must
// set this header (e.g. via a Lambda authorizer or an upstream auth-gateway).
const tenantHeader = "x-anyhook-tenant"

// NewDeliverHandler builds the SQS event-source Lambda handler: one
// invocation per batch. Each record drives one engine delivery attempt via
// SqsTransport.DispatchRecords; a failing record is reported through
// BatchItemFailures so only it is redelivered (requires
// FunctionResponseTypes: ReportBatchItemFailures configured on the event
// source mapping). Mirrors TS createDeliverHandler.
func NewDeliverHandler(env Env, opts AdapterOptions) func(ctx context.Context, event events.SQSEvent) (events.SQSEventResponse, error) {
	engine, adapter := BuildEngine(env, opts)
	return func(ctx context.Context, event events.SQSEvent) (events.SQSEventResponse, error) {
		return adapter.Transport.DispatchRecords(ctx, event, engine.ProcessMessage)
	}
}

// NewSweeperHandler builds the scheduled (EventBridge, fixed 60s rate per
// D2/ADR-0001) Lambda handler: sweeps due SCHED# rows and enqueues them to
// SQS. Mirrors TS createSweeperHandler.
func NewSweeperHandler(env Env) func(ctx context.Context) error {
	return func(ctx context.Context) error {
		_, err := RunSweeper(ctx, env, timeNowMs())
		return err
	}
}

// ingestRequest is the JSON body NewIngestHandler expects.
type ingestRequest struct {
	Type           string          `json:"type"`
	Payload        json.RawMessage `json:"payload"`
	IdempotencyKey string          `json:"idempotencyKey"`
}

// NewIngestHandler builds the API Gateway (HTTP API, payload format 2.0)
// ingest Lambda handler: parses the JSON request body and calls Engine.Send.
//
// The TS ingest handler (handlers.ts) mounts core's framework-agnostic
// createPortalRouter, which also serves endpoint CRUD and the delivery-log
// query surface; that portal router has not been ported to go/core (only the
// engine's own Send/ProcessMessage surface has), so this handler covers just
// the send path -- the one every producer integration needs -- rather than
// the full portal. See the task report for this scope note.
func NewIngestHandler(env Env, opts AdapterOptions) func(ctx context.Context, req events.APIGatewayV2HTTPRequest) (events.APIGatewayV2HTTPResponse, error) {
	engine, _ := BuildEngine(env, opts)
	return func(ctx context.Context, req events.APIGatewayV2HTTPRequest) (events.APIGatewayV2HTTPResponse, error) {
		tenant := req.Headers[tenantHeader]
		if tenant == "" {
			return jsonResponse(401, map[string]string{"error": "missing " + tenantHeader + " header"}), nil
		}

		raw := []byte(req.Body)
		if req.IsBase64Encoded {
			decoded, err := base64.StdEncoding.DecodeString(req.Body)
			if err != nil {
				return jsonResponse(400, map[string]string{"error": "invalid base64 body"}), nil
			}
			raw = decoded
		}

		var body ingestRequest
		if err := json.Unmarshal(raw, &body); err != nil {
			return jsonResponse(400, map[string]string{"error": "invalid JSON body"}), nil
		}

		receipt, err := engine.Send(ctx, core.SendEvent{
			Type: body.Type, Tenant: tenant, Payload: body.Payload, IdempotencyKey: body.IdempotencyKey,
		})
		if err != nil {
			return jsonResponse(400, map[string]string{"error": err.Error()}), nil
		}
		return jsonResponse(202, receipt), nil
	}
}

func jsonResponse(status int, v any) events.APIGatewayV2HTTPResponse {
	body, err := json.Marshal(v)
	if err != nil {
		return events.APIGatewayV2HTTPResponse{StatusCode: 500, Body: fmt.Sprintf(`{"error":%q}`, err.Error())}
	}
	return events.APIGatewayV2HTTPResponse{
		StatusCode: status,
		Headers:    map[string]string{"content-type": "application/json"},
		Body:       string(body),
	}
}

File: packages/aws/src/handlers.ts
/** @fileoverview Lambda entrypoints: API Gateway ingest (portal), SQS delivery consumer, scheduled sweeper. @module @anyhook/aws */
import type { APIGatewayProxyEventV2, APIGatewayProxyResultV2, SQSEvent, SQSBatchResponse, ScheduledEvent } from 'aws-lambda';
import { createPortalRouter } from '@anyhook/core';
import { buildEngine } from './adapter.js';
import { runSweeper } from './sweeper.js';
import { resolveEnv, type AwsConfig } from './config.js';

/**
 * API Gateway (HTTP API, payload format 2.0) → core's framework-agnostic portal router. Like the
 * Cloudflare Worker template, anyhook does NOT authenticate end customers here (G8): mount this
 * behind the producer's own auth, which must set the trusted `x-anyhook-tenant` header (e.g. via a
 * Lambda authorizer or an upstream auth-gateway). See docs/composition-auth.md.
 */
export function createIngestHandler(config: AwsConfig) {
  return async function ingestHandler(event: APIGatewayProxyEventV2): Promise<APIGatewayProxyResultV2> {
    const { engine } = buildEngine(config);
    const router = createPortalRouter(engine);
    const res = await router(toWebRequest(event));
    return toApiGatewayResult(res);
  };
}

/**
 * SQS event-source Lambda: one invocation per batch. Each record drives one engine delivery
 * attempt via `transport.dispatchRecords`; a throwing record is reported through
 * `batchItemFailures` so only it is redelivered (requires `FunctionResponseTypes:
 * ReportBatchItemFailures` configured on the event source mapping).
 */
export function createDeliverHandler(config: AwsConfig) {
  return async function deliverHandler(event: SQSEvent): Promise<SQSBatchResponse> {
    const { engine, adapter } = buildEngine(config);
    await adapter.transport.subscribe((m) => engine.processMessage(m));
    return adapter.transport.dispatchRecords(event);
  };
}

/**
 * Scheduled (EventBridge, fixed 60s rate per D2/ADR-0001) Lambda: sweeps due `SCHED#` rows and
 * enqueues them to SQS. `_event` is accepted for handler-signature parity with the other entrypoints
 * but unused — the sweeper reads only the clock and the table.
 */
export function createSweeperHandler(config: AwsConfig) {
  return async function sweeperHandler(_event: ScheduledEvent): Promise<void> {
    const env = resolveEnv(config);
    await runSweeper(env);
  };
}

// ---- API Gateway (HTTP API v2) <-> Web Request/Response adapters ----

function toWebRequest(event: APIGatewayProxyEventV2): Request {
  const proto = event.headers?.['x-forwarded-proto'] ?? 'https';
  const host = event.headers?.host ?? event.requestContext?.domainName ?? 'localhost';
  const qs = event.rawQueryString ? `?${event.rawQueryString}` : '';
  const url = `${proto}://${host}${event.rawPath}${qs}`;

  const headers = new Headers();
  for (const [k, v] of Object.entries(event.headers ?? {})) {
    if (v !== undefined) headers.set(k, v);
  }

  const method = event.requestContext.http.method;
  const hasBody = method !== 'GET' && method !== 'HEAD' && event.body !== undefined;
  const body = hasBody ? (event.isBase64Encoded ? Buffer.from(event.body!, 'base64') : event.body) : undefined;

  return new Request(url, { method, headers, body });
}

async function toApiGatewayResult(res: Response): Promise<APIGatewayProxyResultV2> {
  const headers: Record<string, string> = {};
  res.headers.forEach((v, k) => {
    headers[k] = v;
  });
  const body = await res.text();
  return { statusCode: res.status, headers, body };
}
